Storybook input fields

BartDeCaluwe opened the pull request:
This PR will add:

- Input component with the following variants: default, error, display, disabled
- Label component that wraps the input field
- Input component stories
- Input snapshot tests

BartDeCaluwe committed:
add: Tailwind css forms plugin

BartDeCaluwe committed:
add: Label component

BartDeCaluwe committed:
add: input component

BartDeCaluwe committed:
add: input component stories

BartDeCaluwe committed:
update: input placeholder prop to be optional

BartDeCaluwe committed:
add: input component snapshot tests

BartDeCaluwe committed:
feat: remove size specific styling

The design docs only specify one size

BartDeCaluwe committed:
feat: use a context to pass validation state from label to input

AntDesign uses css inheritance to style input feedback. This way they can set validation state on the label that affects the input style. Because we use Tailwind css we needed a way to access the label state for our styling.

BartDeCaluwe committed:
style: input label font size and spacing

BartDeCaluwe committed:
add: story decorator to description story

BartDeCaluwe committed:
test: update snapshots

BartDeCaluwe committed:
add: disabled prop to input context

BartDeCaluwe committed:
feat: validation context should not be "error" by default

BartDeCaluwe committed:
formatting

BartDeCaluwe committed:
cleanup: unused variable modifier

BartDeCaluwe committed:
chore: update lock file

BartDeCaluwe committed:
test: update input snapshots

BouweCeunen commented:
It seems that the `label` doesn't change on storybook when I edit it in the Control, is this intended? it does work with the placeholder.

BartDeCaluwe commented:
> It seems that the `label` doesn't change on storybook when I edit it in the Control, is this intended? it does work with the placeholder.

Yeah it has nothing to do with the component but more with how Storybook works.
The "story" params are passed through the Input component and not the Label component. 
But I'll take a look to see if I can fix that.

BartDeCaluwe committed:
feat: make all input story params functional

RuvendeGroote commented:
![image](https://user-images.githubusercontent.com/14349484/154071819-7d40db65-f409-4163-84d2-237efb09ddbc.png)
it's seems like something got broken
https://2468cf26.credix-app-v2.pages.dev/?path=/story/button--primary

RuvendeGroote commented:
if it's actually broken: shouldn't our snapshot tests pick this up?

BartDeCaluwe commented:
> if it's actually broken: shouldn't our snapshot tests pick this up?

It depends on what's actually wrong. Snapshot tests check if the "markup" of your component has changed. If for example I break some css class (which is what I'm guessing has happened) then the markup hasn't changed but the look of the component has.

One solution for this would be to use [Chromatic](https://www.chromatic.com/) (made by the same company behind storybook) which basically compares two screenshots of a component and diffs them. 
We could definitely look into this as I believe it's free for OSS.

BartDeCaluwe committed:
fix: button background color

RuvendeGroote commented:
> Member

alright, for something visual like a component library I think this is a good idea. is chromatic your preferred choice or just something you looked up right now? I (or you if you want) will add a ticket to set this up

BartDeCaluwe commented:
> alright, for something visual like a component library I think this is a good idea. is chromatic your preferred choice or just something you looked up right now? I (or you if you want) will add a ticket to set this up

It's not something I've researched, and I've never used something similar before. It's just what Storybook recommends, as they integrate nicely. 
> "Chromatic tests map 1-to-1 with your stories. Whenever you update a story your tests are synced. This means visual tests are painless to keep up to date."

As per their docs. It would surprise me if there is an alternative that would integrate that well, but I can definitely look into it. (I'll make the ticket)

RuvendeGroote commented on __tests__/__mocks__/matchMedia.mock.js:
```diff
@@ -0,0 +1,13 @@
+Object.defineProperty(window, "matchMedia", {
```
why are we doing this?

BartDeCaluwe replied:
AntDesign uses `matchMedia` for what I assume is responsiveness reasons (as `matchMedia` returns a `MediaQueryList`). But this is something that is not supported in JSDOM (which is the DOM implementation that Jest uses). This causes an error.
Jest has an official "fix" which is what that code does.
More can be read [here](https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom)

BartDeCaluwe committed:
refactor: remove the label component

Having the separate Label component required us to create an input context to pass along the validation state from the label component to its children. This made the component needlessly complex, so we opted for integrating the Label component into the Input component.
